gcloud/storage/batch.py

```diff
@@ -0,0 +1,208 @@
+
+See: https://cloud.google.com/storage/docs/json_api/v1/how-tos/batch
+"""
+from email.encoders import encode_noop
+from email.generator import Generator
+from email.mime.application import MIMEApplication
+from email.mime.multipart import MIMEMultipart
+from email.parser import Parser
+import io
+import json
+import sys
+
+import six
+
+from gcloud._localstack import _LocalStack
+
+
+_BATCHES = _LocalStack()
+
+_PROXIED_ATTRS = [
+    '_make_request',
+    'api_request',
+    'build_api_url',
+    'get_all_buckets',
+    'get_bucket',
+    'create_bucket',
+    'delete_bucket',
+]
+
+
+class MIMEApplicationHTTP(MIMEApplication):
+    """MIME type for ``application/http``.
+
+    Constructs payload from headers and body
+
+    :type headers:  dict
+    :param headers: HTTP headers
+
+    :type body: text or None
+    :param body: HTTP payload
+    """
+    def __init__(self, method, uri, headers, body):
+        if isinstance(body, dict):
+            body = json.dumps(body)
+            headers['Content-Type'] = 'application/json'
+            headers['Content-Length'] = len(body)
+        if body is None:
+            body = ''
+        lines = ['%s %s HTTP/1.1' % (method, uri)]
+        lines.extend(['%s: %s' % (key, value)
+                      for key, value in sorted(headers.items())])
+        lines.append('')
+        lines.append(body)
+        payload = '\r\n'.join(lines)
+        if sys.version_info[0] < 3:  # pragma: NO COVER  Python2
+            MIMEApplication.__init__(self, payload, 'http', encode_noop)
+        else:                        # pragma: NO COVER  Python3
+            super_init = super(MIMEApplicationHTTP, self).__init__
+            super_init(payload, 'http', encode_noop)
+
+
+class Batch(object):
+    """Proxy an underlying connection, batching up change operations.
+
+    :type connection: :class:`gcloud.storage.connection.Connection`
+    :param connection: the connection for which the batch proxies.
+    """
+    def __init__(self, connection):
+        self._connection = connection
+        self._http = _FauxHTTP(connection)
+        self._requests = self._responses = ()
+        for attr in _PROXIED_ATTRS:
+            setattr(self, attr, getattr(connection, attr))
+
+    def finish(self):
+        """Submit a single `multipart/mixed` request w/ deferred requests.
+
+        :rtype: list of tuples
+        :returns: one ``(status, reason, payload)`` tuple per deferred request.
+        :raises: ValueError if no requests have been deferred.
+        """
+        deferred = self._requests = self._http.finalize()
+
+        if len(deferred) == 0:
+            raise ValueError("No deferred requests")
+
+        multi = MIMEMultipart()
+
+        for method, uri, headers, body in deferred:
+            sub = MIMEApplicationHTTP(method, uri, headers, body)
+            multi.attach(sub)
+
+        # Flatten payload
+        if six.PY3:             # pragma: NO COVER  Python3
+            buf = io.StringIO()
```

Comment:
Thanks for the comment. This makes me sad. Isn't `email` in the stdlib? Shouldn't it be good at the boundary of 2 and 3. Le sigh.